src/test/com/fulcrologic/fulcro/algorithms/normalized_state_helpers_spec.cljc

```diff
@@ -111,58 +112,91 @@
       (assertions
         (nsh/get-in state [:person/id 3 :person/spouse :person/name] "not-found") => "not-found")))
 
-
   )
 
 
 ;;============================================================================
 
 (specification "remove-entity*"
-  (behavior "1"
-    (let [state {:a [:b 1]
-                 :b {1 {:c [:d 1]}}
-                 :d {1 {:value 42}}}]
-      (assertions
-        (nsh/get-in state [:a :c :value]) => 42)))
-  )
-
+  (let [denorm-data {:a [[:b 1]] :b [:b 1]}
+        state       {:a      [:b 1]
+                     :top-tm [[:d 1] [:b 1]]
+                     :denorm {:level-1 {:level-2 denorm-data}}
+                     :b      {1 {:c [:d 1] :e [[:b 1] [:d 1]]}}
+                     :d      {1 {:value 42}}}]
+    (assertions
+      "removes the entity itself from the database"
+      (-> (nsh/remove-entity* state [:b 1])
+        (get-in [:b 1])
+        nil?) => true
+      "Removes top-level to-one references"
+      (-> (nsh/remove-entity* state [:b 1]) :a nil?) => true
+      "Removes table-nested to-one references"
+      (-> (nsh/remove-entity* state [:d 1]) (get-in [:b 1 :c]) nil?) => true
+      "Removes top-level to-many refs"
+      (-> (nsh/remove-entity* state [:d 1]) :top-tm) => [[:b 1]]
+      "Removes table-nested to-many refs"
+      (-> (nsh/remove-entity* state [:d 1]) (get-in [:b 1 :c]) nil?) => true
+      "Ignores denormalized data"
+      (-> (nsh/remove-entity* state [:b 1]) (get-in [:denorm :level-1 :level-2])) => denorm-data)))
 
 ;============================================================================
 
 (specification "remove-edge*"
-  (behavior "1"
+  (component "Simple to-one edge removal"
     (let [state {:a [:b 1]
                  :b {1 {:c [:d 1]}}
                  :d {1 {:value 42}}}]
-      #_(assertions
-          (nsh/get-in state [:a :c :value]) => 42)))
-  )
+      (assertions
+        "Removes to-one edge"
+        (nsh/remove-edge* state [:b 1 :c]) => {:a [:b 1]
+                                               :b {1 {}}
+                                               :d {1 {:value 42}}}
+        "Refuses to remove something that is not a normalized edge"
+        (nsh/remove-edge* state [:d 1 :value]) => state
+        (nsh/remove-edge* state [:d 1]) => state
+        (nsh/remove-edge* state [:d]) => state)))
+  (component "Recursive to-one edge removal"
+    (let [state {:a [:b 1]
+                 :b {1 {:c [:d 1]}}
+                 :d {1 {:value 42 :other [:e 2] :keeper [:e 3]}}
+                 :e {2 {:x 1}
+                     3 {:y 1}}}]
+      (assertions
+        "Can cascade a delete across named edges."
```

Comment:
I would normally split this into two assertions, since there are two behaviors: taking the `:other` edge through cascade, and leaving the `:keeper` edge (even though the owner is going away).

Reply:
@awkay, could you please shed some more light on the `recursive to-one edge removal` behavior - I get the first behavior and it's fairly easy to implement but I don't have good clarity over the `cascading` options - specifically when is `:keeper` supposed to go away and when not.

Reply:
Quoting from slack for reference 

> The cascade is a set of keywords.  If you’re removing an entity and it contains any one of those keywords as keys, then you want to re-run remove-edge on that edge
> 
> you have a graph, and an entity may be connected to things it owns and things it refers to.  If it is the owner, then you want to remove that nested thing as well…but that nested thing has the same concerns.
> of course, this kind of cascade config would be best declared at a more global level, but this is a low-level util…that feature could be stacked on at a higher abstraction layer later
> e.g. `:person/spouse` is not an ownership relation…you would not delete Sally because you’re deleting her spouse.  But an “engine” in a “car” entity…those probably get deleted together.
> this is just so you can express that in one command.  Now, it might make more sense to make that part of remove-entity* instead…but the recursive nature will need to pass it to this function too
> so, I’d say `(remove-entity* state-map [:person/id 22] #{:person/children :person/spouse})` would remove all of the people related to entity 22 from the db (recursively…removing the children of children, and children’s spouses, etc.) (
> sometimes that could in fact be what you want, depending on the use-case
> so, maybe remove-edge* is NOT cascading…but remove-entity* is?  Now that I think of it, 
> you probably would not use remove-edge* in that alg.
> but I could see someone wanting it in `remove-edge*`
> 
> say you have a UI component pointing to a person via :comp/person…then `(remove-edge* state-map [:comp/id :thing :comp/person] #{:person/spouse :person/children})`

Reply:
Further discussino, 

> 
> Abhinav Sharma:fulcro: 6:33 PM
> So the example I thought of covers many cases and exposed certain questions which I wasn’t asking before. Things like how would we define ownership which really warrants the removal of a child along with the parent and what is only a reference?
> I’ve found a good general abstraction to implement these and possibly other normalized db helper functions later. However, before going further, if possible, I’d like to connect with you and talk this through - let me know when it’s feasible hmm
> 
> tony.kay 9:24 PM
> Ownership is simply defined by the edge’s key name (e.g. :person/children).  Nothing more complex is warranted at this time.
> that’s why cascade is a set of edge names…it’s all you need to know.  If someone wants finer control, then it becomes a more complicated API…I don’t want that. Simple tools. (edited)
> tony.kay 9:30 PM
> Actually, not that I think of it, we could allow cascade to be a predicate function that receives the “current normalized entity” and returns the list of edges to follow (as a set). (edited)
> Abhinav Sharma:fulcro: 9:52 PM
> Hmm, about ownership well I was thinking that in case of cascading it ends up requiring us to eventually cover up quite a few cases like “Joe and Sally, own different cars with same engines” - for which we need to know how many entities actually reference a specific entity. The way I am approaching this, opens up, to-one/many and from-one/many pathways about ownership.
> I’m definitely in favor of re-thinking this functionality, there must be a simpler way to provide the users with sufficient control :slightly_smiling_face:
> About the time, I think our times zones actually differ by 11:30 hours. If you could update your visibility about time in next 3-4 hours, we can connect for sure
> tony.kay 10:07 PM
> no no…I’m not suggesting full-scale GC.  I’m saying IF you KNOW it is safe to cascade, then you can say it…otherwise, don’t.
> You’re missing a critical aspect for real GC: it is a distributed system.  Just because nobody owns it LOCALLY doesn’t mean you actually want to get rid of it. Just because something isn’t on-screen isn’t a good reaason to remove it (and disconnected entities in the graph often meet this criteria).
> :+1:
> 1
> 
> I say cascade a spouse and that person is someone else’s child…that is an error, and it is my fault.  You’re right that detecting usages could prevent errors in one direction, though.
> The case I’m trying to cover with the API is a very simple one: You often have a subtree and you KNOW you’re the only one using it.  Being able to just point to that tree is what I want with cascade at this point.
> Let’s get the base algs written and tested…we can add more complex stuff as a later step
> but the problem is that even if I know a subtree needs to go away, one of those entities might be pointing to something that should NOT be followed…so being explicit about safe edges is all cascade is for.
> it’s the edges I (the programmer) KNOW are safe to follow
> if I don’t KNOW that, I don’t list the edge
> so, the implementation is quite simple: If the edge is there (to one OR to many) and it is listed in cascade, delete it ALL (edge and entities pointed to, and what they point to (in cascade) etc.)
> it’s quite a basic (and naive) alg.  Doing more complex things is definitely not what I want.
> GC in Fulcro would be a very hard problem, that would require careful thinking by the user of it, meaning it would not be automatic in any useful way….therefore I never plan to write such a thing in core….also, in 3 years of using the lib, I’ve never ever need it for space reasons.
> :+1:
> 1
> 
> it’s simple enough to clean up after yourself if you’re doing heavy things.